### PR TITLE
Overloading connect with factory

### PR DESCRIPTION
**Pull request by tgriesser**
I had actually considered this before doing the `connectFactory` approach in #185. The issue was that it felt wrong to overload the arguments and require first calling the function to determine its return type.

That said, this is the cleanest implementation I could come up with. It checks if the argument in question has `fn.length` of 0 and also returns a function.

This breaks a few tests, but nothing critical - only on some spies or guarantees an arity-0 function is called a certain amount of times. I can clean those up if this looks alright otherwise.

### CODE REVIEW

**Comment by gaearon** [comment]
Rather than look at arg count, can we determine whether they returned functions instead of objects lazily first time we call them from the constructor?

**Comment by tgriesser** [comment]
Took another pass at it, it's a little more verbose / complicated, but does things lazily as you mentioned, all tests are passing and it avoids calling the mapping functions any more times than necessary. 

If this looks alright I can add some tests.

File: src/components/connect.js [hunk 1]
```diff
@@ -14,65 +14,40 @@ const defaultMergeProps = (stateProps, dispatchProps, parentProps) => ({
   ...dispatchProps
 })
 
+function isFunction(fn) {
```
Comment (gaearon) [inline src/components/connect.js]: I would leave those checks inline.

**Comment by gaearon** [comment]
Yes, this looks good. Please do add tests.

**Comment by tgriesser** [comment]
Updated, rebased, tests added. Also noticed when writing the tests that a final shallow equality check on the final merged props would save a render if there is a custom `mergeProps` supplied and nothing has changed, so I added that as well.

File: test/components/connect.spec.js [hunk 2]
```diff
@@ -1509,5 +1509,108 @@ describe('React', () => {
...
+        <ProviderMock store={store}>
+          <Container />
+        </ProviderMock>
+      )
+
+      store.dispatch({ type: 'test' })
+      expect(updatedCount).toBe(0)
+      expect(memoizedReturnCount).toBe(2)
+    })
+
```
Comment (gaearon) [inline test/components/connect.spec.js]: Nit: mind killing the newline here?

File: src/components/connect.js [hunk 3]
```diff
@@ -18,51 +18,33 @@ function getDisplayName(WrappedComponent) {
...
-  const finalMapDispatchToProps = isPlainObject(mapDispatchToProps) ?
+  const mapState = mapStateToProps || defaultMapStateToProps
+  const mapDispatch = isPlainObject(mapDispatchToProps) ?
     wrapActionCreators(mapDispatchToProps) :
     mapDispatchToProps || defaultMapDispatchToProps
+
   const finalMergeProps = mergeProps || defaultMergeProps
-  const doStatePropsDependOnOwnProps = finalMapStateToProps.length !== 1
-  const doDispatchPropsDependOnOwnProps = finalMapDispatchToProps.length !== 1
+  const checkMergedEquals = finalMergeProps !== defaultMergeProps
```
Comment (gaearon) [inline src/components/connect.js]: Can we add a new test for this optimization so we don't regress?

File: src/components/connect.js [hunk 4]
```diff
@@ -96,8 +78,47 @@ export default function connect(mapStateToProps, mapDispatchToProps, mergeProps,
...
+        const isFactory = typeof mappedState === 'function'
+        this.finalMapStateToProps = isFactory ? mappedState : mapState
+        this.doStatePropsDependOnOwnProps = this.finalMapStateToProps.length !== 1
+        return isFactory ? this.computeStateProps(store, props) : checkStateShape(mappedState)
+      }
+
+      computeDispatchProps(store, props) {
+        if (!this.finalMapDispatchToProps) {
+          return this.configureFinalMapDispatch(store, props)
+        }
```
Comment (gaearon) [inline src/components/connect.js]: Nit: please add a newline here

File: src/components/connect.js [hunk 5]
```diff
@@ -96,8 +78,47 @@ export default function connect(mapStateToProps, mapDispatchToProps, mergeProps,
...
+        const { dispatch } = store
+        const dispatchProps = this.doDispatchPropsDependOnOwnProps ?
+          this.finalMapDispatchToProps(dispatch, props) :
+          this.finalMapDispatchToProps(dispatch)
+        return checkStateShape(dispatchProps, true)
```
Comment (gaearon) [inline src/components/connect.js]: Nit: please, in general, format more sparsely.

```js
if (!this.finalMapDispatchToProps) {
  return this.configureFinalMapDispatch(store, props)
}

const { dispatch } = store
const dispatchProps = this.doDispatchPropsDependOnOwnProps ?
  this.finalMapDispatchToProps(dispatch, props) :
  this.finalMapDispatchToProps(dispatch)

return checkStateShape(dispatchProps, true)
```

File: src/components/connect.js [hunk 6]
```diff
@@ -96,8 +78,47 @@ export default function connect(mapStateToProps, mapDispatchToProps, mergeProps,
...
+      }
+
+      configureFinalMapDispatch(store, props) {
+        const mappedDispatch = mapDispatch(store.dispatch, props)
+        const isFactory = typeof mappedDispatch === 'function'
+        this.finalMapDispatchToProps = isFactory ? mappedDispatch : mapDispatch
+        this.doDispatchPropsDependOnOwnProps = this.finalMapDispatchToProps.length !== 1
+        return isFactory ? this.computeDispatchProps(store, props) : checkStateShape(mappedDispatch, true)
```
Comment (gaearon) [inline src/components/connect.js]: Nit: Would be great to break this ternary into three lines like the rest of them

**Comment by gaearon** [comment]
That’s amazing, thank you so much for your work on this. A few more nits and it’s good to go.

**Comment by tgriesser** [comment]
All set with a test and a bit of formatting polish

**Comment by gaearon** [comment]
Would you be up to contributing a section to [Computing Derived Data](https://github.com/rackt/redux/blob/master/docs/recipes/ComputingDerivedData.md) recipe documenting that? I think a lot of people won’t learn about this useful feature otherwise.

**Comment by tgriesser** [comment]
Sure thing, I'll take a look at doing this a little later.

**Comment by ellbee** [comment]
@tgriesser I have started to update the Reselect docs to reflect this change. If you haven't started on this already, would you like me to carry on and document it?

**Pull request merged**